shim.c

```diff
@@ -1966,10 +1966,18 @@ efi_main (EFI_HANDLE passed_image_handle, EFI_SYSTEM_TABLE *passed_systab)
 die:
 		console_print(L"Something has gone seriously wrong: %s: %r\n",
 			      msgs[msg], efi_status);
+
+		/*
+		 * Provide additional information when the SBAT self check fails
+		 */
+		if ( msg == SBAT_SELF_CHECK ) {
+			console_print(L"\nThe current shim has likely been revoked. " 
```

Comment:
Can we be sure here? "likely" isn't great?

Reply:
The other option is to have the message clearly cover all the cases that could land you here:

1) You are somehow trying to boot an otherwise valid out of date OS image and your shim has been revoked.
2) You are somehow booting an out of date shim (perhaps from the wrong device?)
3) You are somehow trying to boot an old file that was inserted maliciously.
4) You are dealing with a file or EFI variable that was somehow corrupted
5) We ran into an internal error (bug in the shim code)

Although that's enough cases that it may make sense to split them up and handle them appropriately. Certainly verify_sbat_section() could differentiate between EFI_SECURITY_VIOLATION and EFI_ERROR, which would group 1-3 and 4&5.

That wording could probably use some more work, but let me know what you think.

I do want to get something like this merged soon, especially since Windows pre-installed systems are very likely to have SbatLevel pre-applied.

Reply:
@jsetje my plan is then the following:
- differentiate errors in verify_sbat_section()  (thanks for the hint with EFI_ERROR)
- propagate which SBAT rule check failed: Something like:  expected at least level 5  for "shim" got "4"
- Output error messages similar to the ones you suggested depending on the information (sbat is wrong, something else failed)

Reply:
Printing the exact levels may be more than we need here. In the end a customer needs to know which package to update, and the package versions have nothing to do with the sbat label.

Reply:
I now updated the error message handling to differentiate between likely .sbat section or internal errors and when an actual revocation happens.

As I had most of the code for showing the component name and expected and found SBAT levels already done, I also now included it in the error message shown to the user.

Error message for revocation looks like this:
![revocation](https://github.com/user-attachments/assets/b3def4b5-1505-4477-ba78-4728a13408f4)

Error message for everything else (in this case I just set the .sbat section to nothing):
![error](https://github.com/user-attachments/assets/a56339d7-cb50-4c84-b020-04a3533e3bf4)

I think this gives users and also distros enough info if something happens.

Reply:
On the one hand I don't love "Something has gone seriously wrong", on the other hand it does convey that we did not want anyone to end up in that situation and it's worth escalating so that we can fix anything that may have automatically put it there, such as a dependency check before applying a revocation. While it made headlines recently, I didn't actually see complaints about that string, just that there wasn't additional useful information.

Reply:
There is two types of information that we want to convey: what went wrong (e.g. shim was revoked via SBAT) and what to do for the user.

The first one should contain enough information for us or the distributions to see what is wrong and the second something that a user can easily understand. 

We already print a decent amount of technical error messages, that I wouldn't want to hide, because those are helpful to debug issues. Maybe something like this:

```
[...]
Something has gone seriously wrong: [...]
SBAT self-check failed! Expected "shim,5" or higher. Got :"shim,4".

Your shim has been revoked via the UEFI Secure Boot SBAT policy on your machine. Please update to a compliant version.

[Vendor custom string here]
```

We can add the option for distributions to include some information where to redirect people, this is then something we would need to audit during reviews and have policies around. @jsetje what do you think? Also wouldn't direct users directly to upstream as most errors should go through the distribution first.

Reply:
We can add a vendor string option at build time quite readily here - is that all we need to do to move this forward? I'm happy with this last suggesrtion from @aronowski here.